packages/block-library/src/query/edit/inspector-controls/taxonomy-controls.js

```diff
@@ -51,29 +52,65 @@ export function TaxonomyControls( { onChange, query } ) {
 	const { postType, taxQuery } = query;
 
 	const taxonomies = useTaxonomies( postType );
-	if ( ! taxonomies || taxonomies.length === 0 ) {
+	if ( ! taxonomies?.length ) {
 		return null;
 	}
 
 	return (
 		<VStack spacing={ 4 }>
 			{ taxonomies.map( ( taxonomy ) => {
-				const termIds = taxQuery?.[ taxonomy.slug ] || [];
-				const handleChange = ( newTermIds ) =>
+				const includeTermIds =
+					taxQuery?.include?.[ taxonomy.slug ] || [];
+				const excludeTermIds =
+					taxQuery?.exclude?.[ taxonomy.slug ] || [];
+				const onChangeTaxQuery = ( newTermIds, key ) => {
+					const newPartialTaxQuery = {
+						...taxQuery?.[ key ],
+						[ taxonomy.slug ]: newTermIds,
+					};
+					// Remove empty arrays from the partial `taxQuery` (include|exclude).
+					if ( ! newTermIds.length ) {
+						delete newPartialTaxQuery[ taxonomy.slug ];
+					}
+					const newTaxQuery = {
+						...taxQuery,
+						[ key ]: !! Object.keys( newPartialTaxQuery ).length
+							? newPartialTaxQuery
+							: undefined,
+					};
 					onChange( {
-						taxQuery: {
-							...taxQuery,
-							[ taxonomy.slug ]: newTermIds,
-						},
+						// Clean up `taxQuery` if all filters are removed.
+						taxQuery: Object.values( newTaxQuery ).every(
+							( value ) => ! value
+						)
+							? undefined
+							: newTaxQuery,
 					} );
-
+				};
 				return (
-					<TaxonomyItem
-						key={ taxonomy.slug }
-						taxonomy={ taxonomy }
-						termIds={ termIds }
-						onChange={ handleChange }
-					/>
+					<Fragment key={ taxonomy.slug }>
+						<TaxonomyItem
+							taxonomy={ taxonomy }
+							termIds={ includeTermIds }
+							oppositeTermIds={ excludeTermIds }
```

Comment:
`oppositeTermIds` is a clever addition, but keep in mind that it _could_ backfire in edge cases where users are changing these filters (from excl to incl or vice versa) by first adding in one field and then removing in the other.

Ultimately, it's your call. :) 

It could also be something explored later, including alternatives like using in-place validation to warn users if they have the same term in both inputs at the same time.

Reply:
Yeah, I'd lean on leaving it as is for now and see if there is any feedback in the future. Personally I thought it's better to do that, because adding a term in both include and exclude will return nothing, that could be confusing too.